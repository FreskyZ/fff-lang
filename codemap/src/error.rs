///! fff-lang
///!
///! codemap/error
///! codemap has its own error type because messages based on codemap for locating source code string

use std::fmt;
use std::cmp;
use std::io::Error as IOError;
use std::path::PathBuf;

pub enum CodeMapError {
    CannotOpenFile(PathBuf, IOError),
    CannotReadFile(PathBuf, IOError),
}
impl fmt::Debug for CodeMapError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            &CodeMapError::CannotOpenFile(ref file_name, ref ioe) => write!(f, "Cannot open file: {}: {}", file_name.display(), ioe),
            &CodeMapError::CannotReadFile(ref file_name, ref ioe) => write!(f, "Cannot read file: {}: {}", file_name.display(), ioe),
        }
    }
}
impl cmp::PartialEq for CodeMapError {
    fn eq(&self, rhs: &CodeMapError) -> bool {
        match (self, rhs) {
            (&CodeMapError::CannotOpenFile(ref name1, ref e1), &CodeMapError::CannotOpenFile(ref name2, ref e2)) => 
                name1 == name2 && format!("{:?}", e1) == format!("{:?}", e2),
            (&CodeMapError::CannotReadFile(ref name1, ref e1), &CodeMapError::CannotReadFile(ref name2, ref e2)) => 
                name1 == name2 && format!("{:?}", e1) == format!("{:?}", e2),
            _ => false,
        }
    }
}
impl cmp::Eq for CodeMapError {
}