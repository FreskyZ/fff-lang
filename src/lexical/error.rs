
use std::io;
pub enum Error {

    CannotOpenFile { file_name: String, e: io::Error },
    CannotReadFile { file_name: String, e: io::Error },
}

use std::fmt;
impl fmt::Debug for Error {

    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use self::Error::*;
        match *self {
            CannotOpenFile { ref file_name, ref e } => {
                write!(f, "Cannot open input file `{}`: {}", file_name, e)
            }
            CannotReadFile { ref file_name, ref e } => {
                write!(f, "Cannot read input file `{}`: {}", file_name, e)
            }
        }
    }
}

impl_display_by_debug!(Error);