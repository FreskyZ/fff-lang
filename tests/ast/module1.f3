
use log::new_logger;
use log::error as log_error;
use log::info as log_info;

// very poor whne no lambda
fn m2(a: i32) -> i32 {
    return a * 2;
}

fn main() {
    assert_eq(itertools::map([2, 3, 4], m2), [4, 6, 8]);
}