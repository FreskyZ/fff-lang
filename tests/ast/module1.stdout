module tests/ast/module1.f3
  use-stmt <2:1-2:20>
    name <2:5-2:19>
      name-segment <2:5-2:7> log
      name-segment <2:10-2:19> new_logger
  use-stmt <3:1-3:28> alias log_error <3:19-3:27>
    name <3:5-3:14>
      name-segment <3:5-3:7> log
      name-segment <3:10-3:14> error
  use-stmt <4:1-4:26> alias log_info <4:18-4:25>
    name <4:5-4:13>
      name-segment <4:5-4:7> log
      name-segment <4:10-4:13> info
  fn <7:1-9:1> m2 <7:4-7:5> () <7:6-7:13>
    fn-param <7:7-7:12> a <7:7-7:7>
      primitive-type i32 <7:10-7:12>
    primitive-type i32 <7:18-7:20>
    block <7:22-9:1>
      ret-stmt <8:5-8:17>
        binary-expr <8:12-8:16> * <8:14-8:14>
          name <8:12-8:12>
            name-segment <8:12-8:12> a
          literal i32 2 <8:16-8:16>
  fn <11:1-13:1> main <11:4-11:7> () <11:8-11:9>
    block <11:11-13:1>
      simple-expr-stmt <12:5-12:56>
        fn-call <12:5-12:55> () <12:14-12:55>
          name <12:5-12:13>
            name-segment <12:5-12:13> assert_eq
          fn-call <12:15-12:43> () <12:29-12:43>
            name <12:15-12:28>
              name-segment <12:15-12:23> itertools
              name-segment <12:26-12:28> map
            array-def <12:30-12:38>
              literal i32 2 <12:31-12:31>
              literal i32 3 <12:34-12:34>
              literal i32 4 <12:37-12:37>
            name <12:41-12:42>
              name-segment <12:41-12:42> m2
          array-def <12:46-12:54>
            literal i32 4 <12:47-12:47>
            literal i32 6 <12:50-12:50>
            literal i32 8 <12:53-12:53>
